File: Week_06/Practicum/Binary Tree Right Side View.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

def bfs(root):
    q = deque([root])
    
    ans = []

    while q:
        ans.append(q[-1].val)
        for _ in range(len(q)):
            current = q.popleft()
        
            if current.left:
                q.append(current.left)
            if current.right:
                q.append(current.right)
    
    return ans

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        return bfs(root) if root else []

File: Week_06/Practicum/Binary Tree Right Side View.cpp
class Solution {
    void dfs(TreeNode* node, int depth, vector<int>& view) {
        if (!node) {
            return;
        }

        if (view.size() == depth) {
            view.push_back(node->val);
        }

        dfs(node->right, depth + 1, view);
        dfs(node->left, depth + 1, view);
    }
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> view;
        dfs(root, 0, view);
        return view;
    }
};
